File: platform/platform-core/fastify/src/lifecycles/06_cors.lifecycle.mjs
/**
 * CORS Configuration Lifecycle Module
 *
 * Enables Cross-Origin Resource Sharing for the Fastify server.
 * This allows the frontend app served from a different port to make API requests.
 */

import cors from '@fastify/cors';

/**
 * Configure CORS on startup (before server.listen).
 * @param {import('fastify').FastifyInstance} server
 * @param {object} config
 */
export async function onStartup(server, config) {
  server.log.info('[lifecycle:cors] Configuring CORS middleware...');

  try {
    // Read CORS settings from config (loaded by 01-app-yaml.lifecycle.mjs)
    server.log.info('[lifecycle:cors] Reading CORS settings from config');
    const origins = server.config?.getNested?.(['cors', 'origins']) || [];
    const originDefault = server.config?.getNested?.(['fastify', 'cors', 'origin']) ?? false;
    const methods = server.config?.getNested?.(['cors', 'methods']) || ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'];
    const allowedHeaders = server.config?.getNested?.(['cors', 'allowedHeaders']) || ['Content-Type', 'Authorization'];
    const credentials = server.config?.getNested?.(['cors', 'credentials']) ?? true;
    const maxAge = server.config?.getNested?.(['cors', 'maxAge']) ?? 86400;

    server.log.debug({ origins, originDefault, methods, allowedHeaders, credentials, maxAge }, '[lifecycle:cors] CORS config values');

    // Use explicit origins list if available, otherwise fall back to origin default
    const origin = origins.length > 0 ? origins : originDefault;

    if (!origin || (Array.isArray(origin) && origin.length === 0)) {
      throw new Error('[lifecycle:cors] No CORS origins provided in config. Set cors.origins in security.yml');
    }

    server.log.info({ origin }, '[lifecycle:cors] Resolved CORS origins');

    await server.register(cors, {
      origin,
      methods,
      allowedHeaders,
      credentials,
      maxAge,
    });

    server.log.info({ methods, credentials, maxAge }, '[lifecycle:cors] CORS enabled successfully');
  } catch (err) {
    server.log.error({ err, hookName: '06-cors' }, '[lifecycle:cors] CORS configuration failed');
    throw err;
  }
}

File: platform/platform-core/fastapi/platform_core_fastapi/lifecycles/06_cors.lifecycle.py
"""
CORS Configuration Lifecycle Module

Enables Cross-Origin Resource Sharing for the FastAPI server.
This allows the frontend app served from a different port to make API requests.

Uses onInit hook (not onStartup) because middleware must be added before app starts.
"""

import logging
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

logger = logging.getLogger("lifecycle:cors")


def onInit(app: FastAPI, config: dict) -> None:
    """Configure CORS middleware on init (before app starts)."""
    logger.info("Starting cors lifecycle hook...")
    try:
        logger.info("Configuring CORS middleware")

        # Read CORS settings from config (loaded by 01_app_yaml.lifecycle.py)
        origins = []
        methods = ["GET", "POST", "PUT", "DELETE", "PATCH"]
        allowed_headers = ["Content-Type", "Authorization"]
        credentials = True
        max_age = 86400

        if hasattr(app.state, 'config'):
            logger.debug("Reading CORS config from app.state.config")
            origins = app.state.config.get_nested('cors', 'origins', default=[])
            methods = app.state.config.get_nested('cors', 'methods', default=methods)
            allowed_headers = app.state.config.get_nested('cors', 'allowedHeaders', default=allowed_headers)
            credentials = app.state.config.get_nested('cors', 'credentials', default=credentials)
            max_age = app.state.config.get_nested('cors', 'maxAge', default=max_age)
        else:
            logger.warning("app.state.config not available, using CORS defaults")

        if not origins:
            raise ValueError("No CORS origins provided in config. Set cors.origins in security.yml")

        import json
        logger.debug("CORS origins: %s", json.dumps(origins))
        logger.debug("CORS methods: %s", methods)
        logger.debug("CORS allowedHeaders: %s", allowed_headers)
        logger.debug("CORS credentials: %s, maxAge: %s", credentials, max_age)

        app.add_middleware(
            CORSMiddleware,
            allow_origins=origins,
            allow_credentials=credentials,
            allow_methods=methods,
            allow_headers=allowed_headers,
            max_age=max_age,
        )

        logger.info("CORS enabled for %d origins", len(origins))
        logger.info("cors lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("cors lifecycle hook failed: %s", exc, exc_info=True)
        raise
